tests/quilec/cleve-gottesman-tests.lisp

```
@@ -0,0 +1,295 @@
+;;;; tests/quilec/cleve-gottesman-tests.lisp
+;;;;
+;;;; Author: Juan M. Bello-Rivas
+
+(in-package #:cl-quil.quilec-tests)
+
+(setf magicl::*default-tensor-type* '(complex double-float))
+
+(defun pauli-to-program (pauli)
+  "Return parsed program that applies PAULI to the current wavefunction."
+  (quil:with-output-to-quil
+    (loop :for d :from 0
+          :with components := (subseq (cl-quil.clifford::pauli-components pauli) 1)
+          :for l :across components :do
+            (format t "~[I~;X~;Z~;RY(pi)~] ~D~%" l d)
+          :finally (terpri))))
+
+(defun pauli-to-matrix (pauli)
+  "Return matrix corresponding to PAULI (i.e., a Kronecker product of I, X, Z, and Y gates)."
+  (quil:parsed-program-to-logical-matrix (pauli-to-program pauli)))
+
+(defun dist (v1 v2)
+  "Return $\ell_\infty$ distance between vectors V1 and V2."
+  (loop :for v1k :across (magicl::storage v1)
+        :for v2k :across (magicl::storage v2)
+        :maximizing (abs (- v1k v2k))))
+
+(defun dot-product (u v)
+  (loop :for uk :across (magicl::storage u)
+        :for vk :across (magicl::storage v)
+        :sum (* uk (conjugate vk))))
```

It's pretty low-stakes here, but this will silently fail if `u` and `v` have different lengths.